### ISSUE TITLE
[RFC] [DTensor] Support for One-Sided Communication in DTensor

### ISSUE THREAD

**Issue by BenBrock** [pytorch/pytorch#182382]
### 🚀 The feature, motivation and pitch

One-sided communication, which allows GPUs to use one-sided put and get operations to move data to and from other GPUs, is increasingly important as machine learning operations become more fine-grained and irregular.  One-sided communication can allow for greater overlap between communication and computation [[1](#ref-1), [2](#ref-2), [3](#ref-3), [4](#ref-4)], including with dedicated kernels that perform fused communication and computation.  One-sided communication also allows asynchronous fetching of data, which is particularly useful for applications that introduce irregularity or load imbalance such as MoE [[4](#ref-4)].  One-sided algorithms can also make it more straightforward to support a wide variety of partitionings and replication factors [[5](#ref-5), [6](#ref-6)].

PyTorch already has support for one-sided communication through SymmetricMemory.  The proposal here is to allow users to optionally enable DTensor to allocate the `_local_tensor` object, which holds each GPU's local shard, using SymmetricMemory.  This then enables dispatch to optimized one-sided algorithms for matrix multiply and other ops.

There are two potential benefits:

- Once users can create DTensors backed by SymmetricMemory, 1) there is a potential performance boost, since some collective can operate slightly faster with pre-pinned memory, and 2) users can write their own one-sided kernels using DTensor.

- Having SymmetricMemory-based DTensors would allow DTensor's default path to select one-sided implementations for ops when one-sided is optimal.  This could also include kernels that perform fused computation and communication [[1](#ref-1), [2](#ref-2), [3](#ref-3)].

I have created three small PRs 1) enabling DTensors to optionally allocate local tensor shards with SymmetricMemory, 2) adding a Python-side get primitive to SymmetricMemory, and 3) modifying shard propagation so that a specific algorithm can be selected and (optionally) attached to an output sharding, with the desired algorithm executed during dispatch.

I'm hoping to receive some feedback about the general idea of enabling one-sided communication for DTensor as well as the specific mechanism in the PRs.  The algorithm selection mechanism in 3) is very much a draft, and I'd like to feedback on whether this general direction is acceptable and what would need to be done to make sure this can work well with `torch.compile`.

### Motivation: Why One-Sided Communication?
In one-sided communication, a process can use put and get operations to read or write from a region of globally accessible memory on other processes.  These operations are one-sided, meaning that they don't require the remote process to participate in communication, unlike a two-sided send/recv or collective operation.  One-sided operations map cleanly onto low-level hardware capabilities like RDMA for InfiniBand, Ethernet, and other inter-node networks, and host-side DMA transfers and device-side load/store for NVLink, Infinity Fabric, and other intra-node interconnects.

One-sided communication has a number of advantages:

- Low latency and high bandwidth; maps directly to hardware
- Straightforward to overlap put/get operations with computation
- Asynchronous, which helps avoid wasted sync time due to load imbalance of compute or communication (e.g. MoE)

The main _disadvantage_ of one-sided communication is that it gives up the optimizations inherent in collective operations, mainly topology awareness.  One-sided communication works best with a flat, high-performance network (e.g. a DGX box or NVL72).

#### Straightforward Overlap, Asynchrony
The traditional DTensor matrix multiply uses collectives and consists of 1) a collective operation, 2) a local matrix multiply, and 3) a collective operation.  Note that of course the collectives are only performed if necessary for the sharding, and the collective after the local matrix multiply is often delayed by using `Partial` sharding annotation to indicate a dimension is only partially accumulated.  There are also often multiple valid communication patterns that produce different shardings.

However, let's take a sketch of an FSDP-like matrix multiply where we perform an AllGather followed by a local matrix multiply.  (We can consider Y, X, and W to be row partitioned in Y = XW: that is, a 1D device mesh with `placements=[Shard(0)]`.  Below, I take some syntactic liberties like referring to the local tensor of X as `X_shard`.)

```python
# Perform Y = XW
W_gathered = all_gather(W)
Y_shard = X_shard @ W_gathered
```

This has two discrete phases, a communication phase and a computation phase.  First, we AllGather W, then we perform a local multiplication.  There is no overlap between the two.

Let's contrast this with a one-sided implementation.  Here, each process iterates through shards of W, retrieves each shard using a one-sided get operation, then performs a matrix multiplication with it.  If we use an asynchronous get operation, we can overlap the get with the local matrix multiplication.  With a large enough matrix multiply, all of the communication time can be hidden with overlap.

```python
W_buf = W.get_shard_async(0)
for i in range(n_shards):
    W_shard = W_buf.get()

    if i + 1 < n_shards:
        W_buf = W.get_shard_async(i+1)

    X_slice = slice(shard_dim * i, shard_dim * (i + 1))

    Y_shard += X_shard[X_slice] @ W_shard
```

Note that there is no synchronization required, unlike collective-based approaches.  This can help in cases where there is load imbalance, either due to imbalanced computation or communication (e.g., different local batch sizes in inference, hotspot experts in MoE, sparsity, etc.).  This means each process can perform all its work without synchronizing with another process.

One disadvantage here is that instead of performing one big matrix multiply, we perform `n_shards` matrix multiplies with a slice of X.  This is necessary because we're not multiplying by the whole `W_gathered`, but with a shard `W_shard`.  This _can_ result in reduced performance, since the roofline peak may be slower for the smaller GEMM.  However, with a large enough size, the one-sided version tends to be optimal, and it's possible to use a performance model to determine which version will perform better [[4](#ref-4)].

Two-sided algorithms like SUMMA and Cannon's algorithm have a similar structure and also allow for overlap.  However, they require synchronization inside the inner loop.  They also require many variants for different styles of partitionings, while the one-sided algorithms remain largely the same.

#### Kernels with Fused Computation and Communication
Additionally, one-sided communication allows GPU-initiated communication from inside kernels.  This allows for fine-grained overlap, since data transfers and matrix multiplies can be overlapped at the level of a warp or thread block, rather than a bulk kernel launch.  There are a wide variety of strategies available for kernels with fused computation and communication, with significant speedups over collective based approaches [[1](#ref-1), [2](#ref-2), [3](#ref-3)].  Fused kernels work by overlapping communication and computation, similar to the host-initiated approach discussed above, but can achieve higher performance for some problems due to the elimination of multiple kernel launches and associated synchronization, as well as the higher degree of asynchrony possible due to scheduling at the warp or thread block level.  Enabling DTensors backed by symmetric memory as well as creating a path for DTensor's dispatch mechanism to select one-sided or other non-default algorithms would open a path for using fused kernels by default when optimal.

#### Supporting More Shardings
One-sided distributed matrix multiply algorithms can support a wide array of partitionings, including some partitionings not currently supported by DTensor.  Some examples include multiplications between matrices with mixed replication factors and 2D replicated partitionings.  One-sided communication yields itself to a slicing-based communication approach that simply directly retrieves shards that are necessary based on the partitioning.  On flat networks like NVLink, these methods are also theoretically communication optimal [[6](#ref-6), [7](#ref-7)].  This could potentially help DTensor support a wider array of shardings.

### Proposed Implementation
The proposal here is to provide the option to have DTensor allocate local tensor shards using SymmetricMemory.  With local tensor shards allocated in SymmetricMemory, algorithms are then free (but not required) to use one-sided communication for data movement.  I also propose extending DTensor's op dispatch mechanism to support one-sided algorithms.  Currently, sharding propagation only selects an output sharding, and communication is performed implicitly through redistribution.  Instead, we can have sharding propagation optionally select a specific algorithm if it detects that operands are stored in SymmetricMemory and a one-sided algorithm (or any new algorithm we wish to register) will perform better.

I have separated this out into a set of discrete PRs:

1. #182377 - Update DTensor to store local tensor shards using SymmetricMemory, if enabled.

2. #182378 - Add a Python-level `get` primitive to SymmetricMemory, useful for implementing one-sided communication algorithms.  (Orthogonal to 1.)

3. #182379 - Update DTensor's matrix multiply dispatch logic to support additional types of algorithms.  This is a draft and where I would like feedback the most.  Specifically, adds an optional `algorithm` field to the output sharding.  This can be set when operands are stored in SymmetricMemory and one-sided algorithms (or potentially other non-default algorithms) perform better.  During dispatch, the appropriate algorithm is called if the `algorithm` is set.  Otherwise, behavior is unchanged.

4. After the basic infrastructure is set up, there are many optimized algorithms that could be added.

#### SymmetricMemory `mem_pool`
One alternative to modifying DTensor to (optionally) directly allocate local shards with SymmetricMemory would be to require users to use SymmetricMemory's `mem_pool`.  `mem_pool` can be used in a `with` clause to direct all memory allocations to use the SymmetricMemory backend's `mem_pool`.  Example from the SymmetricMemory docs:

```python
    dim = 1024
    w = torch.ones(dim, dim, device=device)
    x = torch.ones(1, dim, device=device)

    mempool = symm_mem.get_mem_pool(device)
    with torch.cuda.use_mem_pool(mempool):
        # y will be in symmetric memory
        y = torch.mm(x, w)
```

Similarly, allocations of local shards inside a DTensor object would be done with SymmetricMemory.  The primary advantage to this approach is it might prevent the need to modify the DTensor class.  There are two disadvantages.  First, _all_ allocations inside the `with` block will use SymmetricMemory, not just the DTensor allocations, which may not be desired.  This might result in memory management issues with the amount of symmetric vs. device memory available for backends like NVSHMEM.  Second, with DTensor, not all GPUs will have local shards with the same dimensions: for example, the last shard in a row may be smaller if the device mesh does not evenly divide the row dimension.  I believe this would violate the semantics of `mem_pool`, since allocation is a collective operation where each GPU must allocate the same size buffer.  Specifically, in the NVSHMEM SymmetricMemory backend, I believe this would result in a collective call to `nvshmem_malloc` with different sizes on each GPU, which is undefined behavior.

#### TorchComms
Another alternative might be to make this a TorchComms-only feature.  This could potentially prevent adding features to SymmetricMemory if we expect it to be replaced.

### References

<a id="ref-1"></a>[1] [[arXiv:2305.06942v2](https://arxiv.org/abs/2305.06942)] Optimizing Distributed ML Communication with Fused Computation-Collective Operations

<a id="ref-2"></a>[2] [[arXiv:2406.06858v1](https://arxiv.org/abs/2406.06858)] FLUX: Fast Software-based Communication Overlap On GPUs Through Kernel Fusion

<a id="ref-3"></a>[3] [[arXiv:2511.12500](https://arxiv.org/abs/2511.12500)] Iris: First-Class Multi-GPU Programming Experience in Triton

<a id="ref-4"></a>[4] [[arXiv:2604.01621](https://arxiv.org/abs/2604.01621)] DWDP: Distributed Weight Data Parallelism for High-Performance LLM Inference on NVL72

<a id="ref-5"></a>[5] https://blog.shi-labs.com/distributed-gemm-88be6a481e2b

<a id="ref-6"></a>[6] [[arXiv:2510.08874](https://arxiv.org/abs/2510.08874)] Slicing Is All You Need: Towards A Universal One-Sided Algorithm for Distributed Matrix Multiplication

<a id="ref-7"></a>[7] https://dl.acm.org/doi/10.1145/3731599.3767484

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @weifengpy @tianyu-l @XilunWu @SherlockNoMad @ppwwyyxx

**Comment by anshul-si** [comment]
@kapilsh @fduwjj could you take a look at this please

**Comment by weifengpy** [comment]
@kwen2501 do you have similar plan? I remember you mentioned something similar

**Comment by kwen2501** [comment]
> The proposal here is to provide the option to have DTensor allocate local tensor shards using SymmetricMemory

Sounds good to me.  IIRC, DTensor's creation API is `DTensor.from_local(tensor)`. As long as user provide the `tensor` from symm_mem, then it should work. 

For DTensors that are created as a result of computation, using MemPool is the current way to go.

**Comment by kwen2501** [comment]
Once the inner tensor is on symm_mem, I would expect DTensor's redistribute op to see a perf improvement in all_reduce, all_gather, and reduce_scatter etc, as those would trigger NCCL's sym_mem kernel implementations.
